### PR TITLE
MM-20581 Migrate 'components/admin_console/team_channel_settings/channel/details' module and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by allenlai18**
#### Summary

commented the code on places i need help.   still struggling on ```make check-types```

#### Ticket Link

https://github.com/mattermost/mattermost-server/issues/13491
https://mattermost.atlassian.net/browse/MM-20581

### CODE REVIEW

File: components/admin_console/team_channel_settings/channel/details/channel_modes.tsx [hunk 1]
```diff
@@ -11,7 +11,14 @@ import AdminPanel from 'components/widgets/admin_console/admin_panel';
 import FormattedMarkdownMessage from 'components/formatted_markdown_message.jsx';
 import LineSwitch from '../../line_switch';
 
-const SyncGroupsToggle = ({isSynced, isPublic, isDefault, onToggle}) => (
+interface Props {
```
Comment (allenlai18) [inline components/admin_console/team_channel_settings/channel/details/channel_modes.tsx]: how would you suggest doing this one.  The 3 functional components have the same Prop

File: components/admin_console/team_channel_settings/channel/details/channel_details.test.tsx [hunk 2]
```diff
@@ -4,19 +4,22 @@
...
-import ChannelDetails from './channel_details.jsx';
+import {Group} from 'mattermost-redux/types/groups';
+import {Channel} from 'mattermost-redux/types/channels';
+
+import ChannelDetails from './channel_details';
 
 describe('admin_console/team_channel_settings/channel/ChannelDetails', () => {
     test('should match snapshot', () => {
-        const groups = [{
+        const groups: Partial<Group>[] = [{
```
Comment (allenlai18) [inline components/admin_console/team_channel_settings/channel/details/channel_details.test.tsx]: I am not sure how to fix these

Comment (allenlai18) [inline components/admin_console/team_channel_settings/channel/details/channel_details.test.tsx]: Type 'Partial<Group>[]' is not assignable to type 'Group[]'.ts(2322) . on line 33

File: components/admin_console/team_channel_settings/channel/details/channel_details.tsx [hunk 3]
```diff
@@ -21,32 +24,44 @@ import {ChannelModes} from './channel_modes';
...
+    team: Team;
+    groups: Group[];
+    totalGroups: number;
+    allGroups: {[gid: string]: Group[]};
+    actions: {
+        getGroups: (channelID: string, q?: string, page?: number, perPage?: number) => ActionFunc;
+        linkGroupSyncable: (groupID: string, syncableID: string, syncableType: SyncableType, patch: SyncablePatch) => ActionFunc;
+        unlinkGroupSyncable: (groupID: string, syncableID: string, syncableType: SyncableType) => ActionFunc;
+        membersMinusGroupMembers: (channelID: string, groupIDs: Array<string>, page?: number, perPage?: number) => ActionFunc;
+        setNavigationBlocked: (blocked: boolean) => null;
```
Comment (allenlai18) [inline components/admin_console/team_channel_settings/channel/details/channel_details.tsx]: not sure about the return type on this. its in admin_actions

Comment (devinbinnie) [reply components/admin_console/team_channel_settings/channel/details/channel_details.tsx]: You can define the type inline here for now, we haven't converted that file so there's no type for it yet.

File: components/admin_console/team_channel_settings/channel/details/channel_modes.tsx (hunk 1 above)
Comment (devinbinnie) [reply components/admin_console/team_channel_settings/channel/details/channel_modes.tsx]: I'm alright with using the same interface for all 3 components, we may as well stay DRY for the time being until the components need to change (if they do).

File: components/admin_console/team_channel_settings/channel/details/channel_details.test.tsx [hunk 4]
```diff
@@ -4,26 +4,30 @@
...
             id: '123',
             display_name: 'DN',
             member_count: 3,
         }];
         const allGroups = {
             123: groups[0],
         };
-        const testChannel = {
+        const testChannel: Partial<Channel> & {team_name: string} = {
```
Comment (allenlai18) [inline components/admin_console/team_channel_settings/channel/details/channel_details.test.tsx]: line 55 ``` Type 'string | undefined' is not assignable to type 'string'.
```

**Comment by allenlai18** [comment]
@devinbinnie  can you review when you get a chance?

**Comment by lindy65** [comment]
@allenlai18 @devinbinnie

Please provide direction for what area in the system console needs testing and confirm that the other area to test is channel settings/details?

Thanks!

**Comment by devinbinnie** [comment]
@lindy65 This is just a TypeScript conversion so there shouldn't be anything specific to test. But it would be helpful to verify existing functionality.

On an Enterprise version of Mattermost, under User Management, click on Channels. Then click on any channel to open its screen. Those components have been converted.

File: components/admin_console/team_channel_settings/channel/details/channel_groups.tsx [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+export const ChannelGroups: React.SFC<ChannelGroupsProps> = (props: ChannelGroupsProps): JSX.Element => {
+    const {onGroupRemoved, onAddCallback, totalGroups, groups, removedGroups, channel, synced, setNewGroupRole} = props;
+    return (
+        <AdminPanel
+            id='channel_groups'
+            titleId={synced ? t('admin.channel_settings.channel_detail.syncedGroupsTitle') : t('admin.channel_settings.channel_detail.groupsTitle')}
+            titleDefault={synced ? 'Synced Groups' : 'Groups'}
+            subtitleId={synced ? t('admin.channel_settings.channel_detail.syncedGroupsDescription') : t('admin.channel_settings.channel_detail.groupsDescription')}
+            subtitleDefault={synced ? 'Add and remove team members based on their group membership on the next scheduled sync.' : 'Group members will be added to the channel based on your sync schedule.'}
+            className='channel_groups'
```
Comment (sudheerDev) [inline components/admin_console/team_channel_settings/channel/details/channel_groups.tsx]: Was this addition of className intended?

File: components/admin_console/team_channel_settings/channel/details/channel_modes.tsx [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        />
+    );
+};
+
+export const ChannelModes: React.SFC<Props> = (props: Props): JSX.Element => {
+    const {isPublic, isSynced, isDefault, onToggle} = props;
+    return (
+        <AdminPanel
+            id='channel_manage'
+            className='channel_manage'
```
Comment (sudheerDev) [inline components/admin_console/team_channel_settings/channel/details/channel_modes.tsx]: Same as above with className.

File: components/admin_console/team_channel_settings/channel/details/channel_profile.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+export const ChannelProfile: React.SFC<ChannelProfileProps> = (props: ChannelProfileProps): JSX.Element => {
+    const {team, channel} = props;
+    return (
+        <AdminPanel
+            id='channel_profile'
+            titleId={t('admin.channel_settings.channel_detail.profileTitle')}
+            titleDefault='Channel Profile'
+            subtitleId={t('admin.channel_settings.channel_detail.profileDescription')}
+            subtitleDefault='Summary of the channel, including the channel name.'
+            className='channel_profile'
```
Comment (sudheerDev) [inline components/admin_console/team_channel_settings/channel/details/channel_profile.tsx]: Same here with the className

Comment (allenlai18) [reply components/admin_console/team_channel_settings/channel/details/channel_profile.tsx]: yes so its a required prop in admin_panel.tsx   should i just change the props to className to be optional?

File: components/admin_console/team_channel_settings/channel/details/channel_groups.tsx (hunk 5 above)
Comment (allenlai18) [reply components/admin_console/team_channel_settings/channel/details/channel_groups.tsx]: yes so its a required prop in admin_panel.tsx should i just change the props to className to be optional?

File: components/admin_console/team_channel_settings/channel/details/channel_profile.tsx (hunk 7 above)
Comment (sudheerDev) [reply components/admin_console/team_channel_settings/channel/details/channel_profile.tsx]: @allenlai18 Yes, it makes sense to change it to optional

**Pull request merged**